src/hbbft_rbc.erl

```diff
@@ -209,40 +193,50 @@ check_completion(Data = #rbc_data{n=N, f=F}, H) ->
     K = 2*F,
 
     Shards = maps:values(maps:get(H, Data#rbc_data.stripes, #{})),
-    case erasure:decode(K, M, Shards) of
-        {ok, Msg} ->
-            %% recompute merkle root H
-            {ok, AllShards} = erasure:encode(K, M, Msg),
-            Merkle = merkerl:new(AllShards, fun merkerl:hash_value/1),
-            MerkleRootHash = merkerl:root_hash(Merkle),
-            case H == MerkleRootHash of
-                true ->
-                    %% root hashes match
-                    %% check if ready already sent
-                    case Data#rbc_data.ready_sent of
-                        true ->
-                            %% Figure2. Bullet6
-                            %% check if we have enough readies and enough echoes
-                            %% N-2F echoes and 2F + 1 readies
-                            case length(maps:get(H, Data#rbc_data.num_echoes, [])) >= M andalso length(maps:get(H, Data#rbc_data.num_readies, [])) >= (2*F + 1) of
-                                true ->
-                                    %% decode V. Done
-                                    {Data#rbc_data{state=done}, {result, Msg}};
-                                false ->
-                                    %% wait for enough echoes and readies?
-                                    {Data#rbc_data{state=waiting, msg=Msg}, ok}
-                            end;
-                        false ->
-                            %% send ready(h)
-                            {Data#rbc_data{state=waiting, msg=Msg, ready_sent=true}, {send, [{multicast, {ready, H}}]}}
+    Echoes = maps:get(H, Data#rbc_data.num_echoes, []),
+    Readies = maps:get(H, Data#rbc_data.num_readies, []),
+
+    case {length(Echoes), length(Readies)} of
+        {NumEchoes, NumReadies} when NumEchoes >= N - 2*F andalso NumReadies >= 2*F + 1 ->
```

Comment:
I'm wondering if the order of `NumEchoes >= N - 2*F` vs `NumReadies >= 2*F + 1` matters here?

Reply:
the order should not matter, at least from how I read the protocol description quoted immediately below

Reply:
Addressed